pkg/trace/agent/normalizer.go

```diff
@@ -41,13 +42,31 @@ var (
 	Year2000NanosecTS = time.Date(2000, time.January, 1, 0, 0, 0, 0, time.UTC).UnixNano()
 )
 
+// fallbackServiceNames holds a map of default service names to use. It maps
+// tracer language to service name.
+var fallbackServiceNames sync.Map
+
+// fallbackService returns the fallback service name for a service
+// belonging to language lang.
+func fallbackService(lang string) string {
+	if lang == "" {
+		return DefaultServiceName
+	}
+	if v, ok := fallbackServiceNames.Load(lang); ok {
+		return v.(string)
+	}
+	var str strings.Builder
+	str.WriteString("unnamed-")
+	str.WriteString(lang)
+	str.WriteString("-service")
+	fallbackServiceNames.Store(lang, str.String())
+	return str.String()
+}
+
 // normalize makes sure a Span is properly initialized and encloses the minimum required info, returning error if it
 // is invalid beyond repair
 func normalize(ts *info.TagStats, s *pb.Span) error {
-	fallbackServiceName := DefaultServiceName
-	if ts.Lang != "" {
-		fallbackServiceName = fmt.Sprintf("unnamed-%s-service", ts.Lang)
-	}
+	defaultService := func() string { return fallbackService(ts.Lang) }
```

Comment:
Do we need `defaultService`? I think just using `fallbackService(ts.Lang)` is clearer, even if repeated.

Reply:
yes also likely less overhead